device_backends/include/NumericAddressedBackendRegisterAccessor.h

```diff
@@ -78,56 +80,136 @@ namespace ChimeraTK {
+          boost::shared_ptr<NumericAddressedLowLevelTransferElement>& rawAccessor, size_t& startAddress,
+          DataConverterType&, const bool& isNotWriteable)
+      : _buffer_2D(buffer), _rawAccessor(rawAccessor), _startAddress(startAddress), _isNotWriteable(isNotWriteable) {}
+
+      void doPostRead();
+      void doPreWrite();
+      void doPostWrite();
+    };
+
     template<typename UserType, typename DataConverterType, bool isRaw>
     void NumericAddressedPrePostActionsImplementor<UserType, DataConverterType, isRaw>::doPostRead() {
-      auto itsrc = _rawAccessor->begin(_startAddress);
-      _dataConverter.template vectorToCooked<UserType>(itsrc, itsrc + _buffer_2D[0].size(), _buffer_2D[0].begin());
+      callForRawType(_registerInfo->getDataDescriptor().rawDataType(), [this](auto t) {
+        typedef decltype(t) RawType;
+        auto itsrc = (RawType*)_rawAccessor->begin(_startAddress);
+        _dataConverter.template vectorToCooked<UserType>(itsrc, itsrc + _buffer_2D[0].size(), _buffer_2D[0].begin());
+      });
     }
 
     template<typename UserType, typename DataConverterType, bool isRaw>
     void NumericAddressedPrePostActionsImplementor<UserType, DataConverterType, isRaw>::doPreWrite() {
       if(_isNotWriteable) {
-        throw ChimeraTK::logic_error("NumericAddressedBackend: "
-                                     "Writeing to a non-writeable register is not allowed"
-                                     /*"(Register name: " + _registerPathName + ")*/ ".");
-      }
-      auto itsrc = _rawAccessor->begin(_startAddress);
-      for(auto itdst = _buffer_2D[0].begin(); itdst != _buffer_2D[0].end(); ++itdst) {
-        *itsrc = _dataConverter.template toRaw<UserType>(*itdst);
-        ++itsrc;
+        throw ChimeraTK::logic_error(
+            "NumericAddressedBackend: Writeing to a non-writeable register is not allowed (Register name: " +
+            _registerInfo->getRegisterName() + ").");
       }
+      callForRawType(_registerInfo->getDataDescriptor().rawDataType(), [this](auto t) {
+        typedef decltype(t) RawType;
+        auto itsrc = (RawType*)_rawAccessor->begin(_startAddress);
+        for(auto itdst = _buffer_2D[0].begin(); itdst != _buffer_2D[0].end(); ++itdst) {
+          *itsrc = _dataConverter.template toRaw<UserType>(*itdst);
+          ++itsrc;
+        }
+      });
     }
 
     // special implementations for int32 raw
     template<typename DataConverterType>
     void NumericAddressedPrePostActionsImplementor<int32_t, DataConverterType, true>::doPostRead() {
-      if(!_rawAccessor->isShared) {
-        _buffer_2D[0].swap(_rawAccessor->rawDataBuffer);
-      }
-      else {
-        auto itsrc = _rawAccessor->begin(_startAddress);
-        auto itdst = _buffer_2D[0].begin();
-        memcpy(&(*itdst), &(*itsrc), _buffer_2D[0].size() * sizeof(int32_t));
-      }
+      auto itsrc = _rawAccessor->begin(_startAddress);
```

Comment:
Basically no. swapping needs both vectors to have the same type. The buffer we are swapping from is the transfer buffer of the low level transfer element. This buffer might be shared across different accessors, which might have a different raw size. (It is not the alignment size which is important but the raw size of the individual register.)

Since the optimisation only works for unshared registers, in theory one might probably find some complicated way to template this to the right type and exchange the buffer with a int8_t-typed one the moment it is being shared. I am pretty sure this will involve an extra indirection for *all* access to this buffer (raw and cooked). Since we are talking about an optimisation which affects raw transfers only - and raw transfers are meant in ChimeraTK as a mere debugging interface e.g. for firmware developers and tools like QtHardMon - such change would probably do more harm than good.

Reply:
I agree that all use cases we have now are for debugging, and the copy is perfectly acceptable.

To go for the high-performance scenario I mentioned it is perfectly fine to get a char raw accessor, because one will do a reinterpret-cast anyway. Maybe one can add a special implementation for int8_t if this is needed. (While it is still debatable if breaking the abstraction is not already calling for using something more low level than DeviceAccess, which is not for thie review).

Reply:
If we some day have such a high performance application, we need to discuss this carefully. I would not consider this "perfectly fine". It may be a last resort solution, if there are no other options and one would have to make even more compromises by not using DeviceAccess. From experience I can tell, that memcpy is often not the performance bottleneck. One would have to have a very particular situation for this to play any role at all. If we are in that position, we can still investigate how to optimise this. At the moment, I would rather try to optimise the cooked data path further, if possible...